model/validation.go

```diff
@@ -204,16 +204,367 @@ func (t Tags) Validate() error {
+func ValidateL4MatchAttributes(ma *proxyconfig.L4MatchAttributes) error {
+	var retVal error
+
+	if ma.SourceSubnet != nil {
+		for _, subnet := range ma.SourceSubnet {
+			if err := validateSubnet(subnet); err != nil {
+				retVal = multierror.Append(retVal, err)
+			}
+		}
+	}
+
+	if ma.DestinationSubnet != nil {
+		for _, subnet := range ma.DestinationSubnet {
+			if err := validateSubnet(subnet); err != nil {
+				retVal = multierror.Append(retVal, err)
+			}
+		}
+	}
+
+	return retVal
+}
+
+func validatePercent(err error, val int32, label string) error {
+	if val > 100 {
+		err = multierror.Append(err, fmt.Errorf("%v must not exceed 100", label))
+	}
+	if val < 0 {
+		err = multierror.Append(err, fmt.Errorf("%v must be in range 0..100", label))
+	}
+	return err
+}
+
+func validateFloatPercent(err error, val float32, label string) error {
+	if val > 100.0 {
+		err = multierror.Append(err, fmt.Errorf("%v must not exceed 100", label))
+	}
+	if val < 0.0 {
+		err = multierror.Append(err, fmt.Errorf("%v must be in range 0..100", label))
+	}
+	return err
+}
+
+// ValidateDestinationWeight validates DestinationWeight
+func ValidateDestinationWeight(dw *proxyconfig.DestinationWeight) error {
+	var retVal error
+
+	if dw.Destination != "" {
+		if err := validateFQDN(dw.Destination); err != nil {
+			retVal = multierror.Append(retVal, err)
+		}
+	}
+
+	// We do not validate tags because they have no explicit rules
+	if err := Validate(dw.Tags); err != nil {
+		retVal = multierror.Append(retVal, err)
+	}
+
+	retVal = validatePercent(retVal, dw.Weight, "weight")
+
+	return retVal
+}
+
+// ValidateHTTPTimeout validates HTTP Timeout
+func ValidateHTTPTimeout(timeout *proxyconfig.HTTPTimeout) error {
+	var retVal error
+
+	if simple := timeout.GetSimpleTimeout(); simple != nil {
+		if simple.TimeoutSeconds < 0 {
+			retVal = multierror.Append(retVal, fmt.Errorf("timeout_seconds must be in range [0..]"))
+		}
+
+		// We ignore override_header_name
+	}
+
+	return retVal
+}
+
+// ValidateHTTPRetries validates HTTP Retries
+func ValidateHTTPRetries(retry *proxyconfig.HTTPRetry) error {
+	var retVal error
+
+	if simple := retry.GetSimpleRetry(); simple != nil {
+		if simple.Attempts < 0 {
+			retVal = multierror.Append(retVal, fmt.Errorf("attempts must be in range [0..]"))
+		}
+
+		// We ignore override_header_name
+	}
+
+	return retVal
+}
+
+// ValidateHTTPFault validates HTTP Fault
+func ValidateHTTPFault(fault *proxyconfig.HTTPFaultInjection) error {
+	var retVal error
+
+	if fault.GetDelay() != nil {
+		if err := validateDelay(fault.GetDelay()); err != nil {
+			retVal = multierror.Append(retVal, err)
+		}
+	}
+
+	if fault.GetAbort() != nil {
+		if err := validateAbort(fault.GetAbort()); err != nil {
+			retVal = multierror.Append(retVal, err)
+		}
+	}
+
+	return retVal
+}
+
+// ValidateL4Fault validates L4 Fault
+func ValidateL4Fault(fault *proxyconfig.L4FaultInjection) error {
+	var retVal error
```

Comment:
I felt the L4Fault itself was valid.  The flaw is that it was put into a route-rule, and it is not implemented there.  The caller throws it:
```
	if value.L4Fault != nil {
		if err := ValidateL4Fault(value.L4Fault); err != nil {
			retVal = multierror.Append(retVal, err)
		}
		retVal = multierror.Append(retVal, fmt.Errorf("L4 faults are not implemented"))
	}
```
In this way it would be possible to write tests for the validity of naked L4 faults but still get the desired behavior for the CLI and the overall model validation.